use-cases/virgo/trainer.py

```diff
@@ -385,7 +379,278 @@ def train(self):
+
+    def create_model_loss_optimizer(self) -> None:
+        # Select generator
+        generator = self.training_config["generator"]
+        scaling = 0.02
+        if generator == "simple":
+            self.model = Decoder(3, norm=False)
+        elif generator == "deep":
+            self.model = Decoder_2d_deep(3)
+        elif generator == "resnet":
+            self.model = GeneratorResNet(3, 12, 1)
+            scaling = 0.01
+        elif generator == "unet":
+            self.model = UNet(input_channels=3, output_channels=1, norm=False)
+        else:
+            raise ValueError("Unrecognized generator type! Got", generator)
+
+        init_weights(self.model, "normal", scaling=scaling)
+
+        # Select loss
+        loss = self.training_config["loss"]
+        if loss == "l1":
+            self.loss = nn.L1Loss()
+        elif loss == "l2":
+            self.loss = nn.MSELoss()
+        else:
+            raise ValueError("Unrecognized loss type! Got", loss)
+
+        # Optimizer
+        self.optimizer = torch.optim.Adam(
+            self.model.parameters(), lr=self.training_config["learning_rate"]
+        )
+
+        # First, define strategy-wise optional configurations
+        if isinstance(self.strategy, RayDeepSpeedStrategy):
+            # Batch size definition is not optional for DeepSpeedStrategy!
+            distribute_kwargs = dict(
+                config_params=dict(
+                    train_micro_batch_size_per_gpu=self.training_config["batch_size"]
+                )
+            )
+        else:
+            distribute_kwargs = {}
+
+        # Distributed model, optimizer, and scheduler
+        self.model, self.optimizer, _ = self.strategy.distributed(
+            self.model, self.optimizer, **distribute_kwargs
+        )
+
+    def create_dataloaders(
+        self,
+        train_dataset: Dataset,
+        validation_dataset: Dataset | None = None,
+        test_dataset: Dataset | None = None,
+    ) -> None:
+        """Override the create_dataloaders function to use the custom_collate function."""
+        # This is the case if a small dataset is used in-memory
+        # - we can use the default collate_fn function
+        if isinstance(train_dataset, TensorDataset):
+            return super().create_dataloaders(
+                train_dataset=train_dataset,
+                validation_dataset=validation_dataset,
+                test_dataset=test_dataset,
+            )
+
+        # If we are using a custom dataset for the large dataset,
+        # we need to overwrite the collate_fn function
+        self.train_dataloader = self.strategy.create_dataloader(
+            dataset=train_dataset,
+            batch_size=self.training_config["batch_size"],
+            num_workers=self.training_config["num_workers_dataloader"],
+            pin_memory=self.training_config["pin_gpu_memory"],
+            # generator=self.torch_rng,
```

Comment:
it would be nice to pass the PRNG for reproducibility

Reply:
Okay added!